orbit/pkg/update/update.go

```diff
@@ -305,6 +414,111 @@ func (u *Updater) Download(repoPath, localPath string) error {
 	return nil
 }
 
+func goosFromPlatform(platform string) (string, error) {
+	switch platform {
+	case "macos", "macos-app":
+		return "darwin", nil
+	case "windows", "linux":
+		return platform, nil
+	default:
+		return "", fmt.Errorf("unknown platform: %s", platform)
+	}
+}
+
+// checkExec checks/verifies a downloaded executable target by executing it.
+func (u *Updater) checkExec(target, tmpPath string) error {
+	localTarget, err := u.localTarget(target)
+	if err != nil {
+		return err
+	}
+	platformGOOS, err := goosFromPlatform(localTarget.info.Platform)
+	if err != nil {
+		return err
+	}
+	if platformGOOS != runtime.GOOS {
+		// Nothing to do, we can't check the executable if running cross-platform.
+		// This generally happens when generating a package from a different platform
+		// than the target package (e.g. generating an MSI package from macOS).
+		return nil
+	}
+
+	if strings.HasSuffix(tmpPath, ".tar.gz") {
+		if err := extractTarGz(tmpPath); err != nil {
+			return fmt.Errorf("extract %q: %w", tmpPath, err)
+		}
+		tmpDirPath := filepath.Join(filepath.Dir(tmpPath), localTarget.info.ExtractedExecSubPath[0])
+		defer os.RemoveAll(tmpDirPath)
+		tmpPath = filepath.Join(append([]string{filepath.Dir(tmpPath)}, localTarget.info.ExtractedExecSubPath...)...)
+	}
+
+	// Note that this would fail for any binary that returns nonzero for --help.
+	out, err := exec.Command(tmpPath, "--help").CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("exec new version: %s: %w", string(out), err)
+	}
+	return nil
+}
+
+// extractTagGz extracts the contents of the provided tar.gz file.
+func extractTarGz(path string) error {
```

Comment:
Random thought: we might want to eventually create a `TargetFile` type that knows how to check itself, extract itself, etc, that works for all file types.